File: inference_framework-java/src/main/java/com/inferenceframework/processing/FeatureStore.java
package com.inferenceframework.processing;  // processing module

import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Manages loading and retrieval of feature vectors for user_ids from JSON file.
 * Uses Jackson for JSON parsing.
 */
public class FeatureStore {
    private static final Logger logger = LoggerFactory.getLogger(FeatureStore.class);

    private final String featuresPath;
    private Map<String, Map<String, Object>> features;

    /**
     * Constructor loads features from JSON.
     */
    public FeatureStore(String featuresPath) {
        this.featuresPath = featuresPath;
        this.features = loadFeatures();
    }

    @SuppressWarnings("unchecked")
    private Map<String, Map<String, Object>> loadFeatures() {
        // Load features from JSON file using Jackson.
        ObjectMapper mapper = new ObjectMapper();
        File file = new File(featuresPath);
        if (!file.exists()) {
            logger.warn("Features file not found: {}. Using empty store.", featuresPath);
            return new HashMap<>();
        }
        try {
            // Load as Map<String, Map> - fix incompatible type from raw Jackson Map<String, Object>
            Map<String, Object> loaded = mapper.readValue(file, Map.class);
            logger.info("Loaded features for {} users.", loaded.size());
            // Safe cast with suppression for generics (JSON deserialized structure matches)
            return (Map<String, Map<String, Object>>) (Map) loaded;
        } catch (IOException e) {
            logger.error("Failed to load features", e);
            return new HashMap<>();
        }
    }

    /**
     * Get feature vector for a user_id.
     */
    public Map<String, Object> getFeatureVector(String userId) {
        return features.getOrDefault(userId, new HashMap<>());
    }

    /**
     * Get features for multiple user_ids (batch).
     */
    public Map<String, Map<String, Object>> getFeaturesForUsers(List<String> userIds) {
        Map<String, Map<String, Object>> result = new HashMap<>();
        for (String uid : userIds) {
            result.put(uid, getFeatureVector(uid));
        }
        return result;
    }

    /**
     * Get all features (for convenience in processing).
     */
    public Map<String, Map<String, Object>> getAllFeatures() {
        return features;
    }
}

File: inference_framework/inference_framework/feature_store.py
import json
from pathlib import Path
from typing import Dict, Any, Optional, List
import logging

logger = logging.getLogger(__name__)

class FeatureStore:
    """Manages loading and retrieval of feature vectors for user_ids from JSON file."""
    
    def __init__(self, features_path: str):
        self.features_path = Path(features_path)
        self.features: Dict[str, Dict[str, Any]] = self._load_features()
    
    def _load_features(self) -> Dict[str, Dict[str, Any]]:
        """Load features from JSON file."""
        if not self.features_path.exists():
            logger.warning(f"Features file not found: {self.features_path}. Using empty store.")
            return {}
        try:
            with open(self.features_path, 'r') as f:
                features = json.load(f)
            logger.info(f"Loaded features for {len(features)} users.")
            return features
        except Exception as e:
            logger.error(f"Failed to load features: {str(e)}")
            return {}
    
    def get_feature_vector(self, user_id: str) -> Optional[Dict[str, Any]]:
        """Get feature vector for a user_id."""
        return self.features.get(user_id)
    
    def get_features_for_users(self, user_ids: List[str]) -> Dict[str, Dict[str, Any]]:
        """Get features for multiple user_ids."""
        return {uid: self.features.get(uid, {}) for uid in user_ids}
